### ISSUE TITLE
Issue with alg._A not updating properly

### ISSUE THREAD

**Issue by mkaguer** [PMEAL/OpenPNM#1548]
I wrote a script where I run transient fickian diffusion algorithm. My problem required that I stop the solver to update models including diffusive_conductance model and then restart solver using last solution as new initial condition. While my diffusive_conductance values were updated properly when I printed alg._A I could see that the A matrix wasn't being updated properly. I worked around this issue by recreating the algorithm object on every iteration. @ma-sadeghi thought that I shouldn't have to recreate the algorithm object for this to work.

**Comment by ma-sadeghi** [comment]
@mkaguer Can you share a short script here that reproduces the bug?

**Comment by mkaguer** [comment]
@magnaou suggested looking into `alg.settings['cache_A']` and make sure it is set to `False`. This fixed my problem. Perhaps we need to consider having `False` as the default setting?

**Comment by ma-sadeghi** [comment]
`cache_A` is not supposed to be exposed to the user, and is our last resort, we need to investigate why A is not updating.

**Comment by ma-sadeghi** [comment]
I looked into this and indeed, it's a bug similar to the one fixed in PR #1511. So, basically, we already fixed this bug for `ReactiveTransport` but forgot to fix it for `TransientReactiveTransport` as well. Essentially, we need to change `cache_A` to `False` if `alg.settings['conductance']` is in `alg._get_iterative_props`. Ideally, this should happen in `GenericTransport._build_A`, but the problem is that `_get_iterative_props` is a special method of `ReactiveTransport` and therefore, cannot be called from `GenericTransport`. I'd propose moving it to `GenericTransport` since iterative properties are not limited to source terms. For instance, concentration-dependent diffusion coefficient makes `throat.diffusive_conductance` an iterative property even in the absence of reactions. So, I'd say this method belongs to `GenericTransport`.

**Comment by jgostick** [comment]
I agree with moving _get_Iterative_props to GenericTransport for the reasons you point out.  There are two other options, which I feel like we should consider: 

1. Create a new class between GenericTransport and ReactiveTransport, called NonLinearTransport (or IterativeTransport, or VariablePropertyTransport), which basically has the iterative property stuff added.
2. Use multiple inheritance, and create a class called VariableProps, which gets inherited by Reactive and TransientReactive, to add the iterative property stuff.

I don't like the first once since the inheritance gets long and hard to follows, but I actually don't hate the second option. Of course just adding it all to GenericTransport is also a fine soln.

**Comment by ma-sadeghi** [comment]
@jgostick Actually, I came up with another solution that requires no change to the code/class layout. I already made a PR (#1571), but here's the new `_build_A` (in `GenericTransport`):

( _the `try` and `except` clause is new_)
```python
    def _build_A(self):
        r"""
        Builds the coefficient matrix based on conductances between pores.
        The conductance to use is specified in the algorithm's ``settings``
        under ``conductance``.  In subclasses (e.g. ``FickianDiffusion``)
        this is set by default, though it can be overwritten.
        """
        gvals = self.settings['conductance']
        if not gvals:
            raise Exception('conductance has not been defined on this algorithm')
        # Decide if caching of A and b is allowed
        try:
            if gvals in self._get_iterative_props():
                self.settings.update({"cache_A": False, "cache_b": False})
        except AttributeError:
            pass
        if not self.settings['cache_A']:
            self._pure_A = None
        if self._pure_A is None:
            network = self.project.network
            phase = self.project.phases()[self.settings['phase']]
            g = phase[gvals]
            am = network.create_adjacency_matrix(weights=g, fmt='coo')
            self._pure_A = spgr.laplacian(am).astype(float)
        self.A = self._pure_A.copy()
```
So, basically, although `GenericTransport` doesn't have access to neither of `_get_iterative_props` and `_get_variable_props`, but since all other subclassed algorithms do, when `_build_A` is initiated from those algorithms, the added `try` and `except` clause indeed works. 

This way, we don't need any adhoc solutions anymore, since `_build_A` is the common step between pretty much all our algorithms.

What do you think?

**Comment by jgostick** [comment]
It's certainly good enough for a quick patch, but a bit hacky, and not self explanatory.  I'd say lets go with this approach for now, but make a note somewhere to consider a proper, explicit inheritance approach.
